### PR TITLE
[jquery] Update type definitions

### LINKED ISSUE #2725: Problem with jQuery Definitions
Hi, I've faced an issue with the Promises definitions on JQuery.

See my code below:

```
function first(): JQueryPromise<string>
{
    return $.when('1');
}

function test()
{
    $.when()
        .then(() =>
        {
            return first();
        })
        .then((value) =>
        {
            var str: string = value; //<--- type mismatch here.
        });
}
```

Typescript is expecting that value is of type JQueryPromise instead of "string".

If I cast value to any, I am able to make it work.

I've posted this on StackOverflow (http://stackoverflow.com/q/25492550/3230040) and was instructed to report it here to be fixed.

Thanks

### LINKED ISSUE #5630: jQuery Deferred
Recently I've found a mismatch in definition of `resolveWith()` method of `JQueryDeferred` class.
According to documentation: https://api.jquery.com/deferred.resolveWith

``` javascript
deferred.resolveWith( context [, args ] ) // args must be an Array
```

Your definition:

``` typescript
resolveWith(context: any, value?: T, ...args: any[]): JQueryDeferred<T>
```

So the correct call of this method causes syntax error:

``` typescript
var def:JQueryDeferred<string>;
var data:string;
def.resolveWith(this, [data]); // <-- syntax error here: TS2345: Argument of type 'string[]' is not assignable to parameter of type 'string'.
def.resolveWith(this, <any>[data]); // <-- this is a workaround to avoid that error
```

`rejectWith()` method also have incorrect definition but it does not produce syntax error.

### LINKED ISSUE #10159: $.when definition expects common result interface
- [Y ] I tried using the latest `xxxx/xxxx.d.ts` file in this repo and had problems.
- [ Y] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ Y] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [ Y] I want to talk about `DefinitelyTyped/jquery.d.ts`.
  - The authors of that type definition are cc/ @borisyankov, @choffmeister, @Steve-Fenton 

I have a $.when in the following file which resolves 3 different promise types:

http://embed.plnkr.co/IfFpZuYLJkL4GkbSdoe7/

The Typescript Definition seems to expect that all Result arguments implement a common interface?  

image

Is there any way around this?

### LINKED ISSUE #17298: @types/jquery 3.2.2
- [x] I tried using the `@types/jquery` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @leonard-thieu @johnnyreilly 

Hi, I upgraded to @types/jquery@3.2.2 after issues with @3.2.1 (described in pull #17276 ). Now I get the following error, when doing $window = jQuery(window):
`TS2322: Type 'JQueryStatic<HTMLElement>' is not assignable to type 'JQuery<HTMLElement>'.
  Property 'jquery' is missing in type 'JQueryStatic<HTMLElement>'.`

If I define the variable as JQueryStatic, I get more errors:
`error TS2339: Property 'on' does not exist on type 'JQueryStatic<HTMLElement>'`

I include jquery as follows:
`import * as jQuery from 'jquery'; ("allowSyntheticDefaultImports": true)`

I'm trying to listen to the window scroll state, ie:
`header.$window.on('scroll', debounce(header.updateScrollState, 16));`

Either we should be able to convert JQueryStatic to JQuery, or have the proper types available on the JQueryStatic?

### LINKED ISSUE #17571: @types/jquery
Invalid typing in JQuery for Deferred and Promise
```
var test = $.Deferred();
var testPromise = test.promise();
var testPromise2 = testPromise.done(() => { });
```
test is type of Deferred<any, any, any> ?
testPromise is type of Promise<any, any, any> ?
and last why is testPromise2  type of **Deferred<any, any, any>** when testPromise is  Promise<any, any, any> ?

This cause failing of build typescript if you have function that should return promise and return deferred object.

### PR DESCRIPTION
**Pull request by leonard-thieu**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
https://jquery.com/upgrade-guide/3.0/
- ~~Increase the version number in the header if appropriate.~~
- ~~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~~

**Summary of changes**
- Closes #17298 (was fixed by #17301)
- Fixes #5630 
- Fixes #2725 
- Fixes #10159 
- Fixes #17571 
- `AjaxSettings`
  - `headers` type changed to `PlainObject<string | null | undefined>` (was `PlainObject`)
  - `xhrFields` now only accepts writable properties on `XMLHttpRequest`
- `StatusCodeCallbacks`
  - Added the rest of the 3xx, 4xx, and 5xx status codes
- `jqXHR`
  - `responseXML` and `responseJSON` are now optional
  - Removed `jqXHR.ProgressCallback`
    - jQuery doesn't send progress updates for jqXHR objects
- `Deferred`
  - Fixed signatures for `notifyWith()`, `rejectWith()`, and `resolveWith()`
  - Add `exceptionHook` property
- `Promise`
  - Fixed return type on methods that returned `this`
  - Added 2- and 3-argument Promise types
- `Callbacks`
  - Now accepts a type parameter representing the type of the callbacks
  - Fixed signatures for `add()`, `fireWith()`, and `remove()`
- `Event`
  - Refactored into its instance, generic, and static components
  - Removed `originalTarget` property
  - `Event(properties)` now requires `type`
- `JQuery`
  - `offset()` can now return `undefined`
  - Index signature is no longer read only
- `JQueryStatic`
  - `isEmptyObject()` is now a user-defined type guard
  - `when()`
    - Updated return types to reflect that the Promise returned does not emit progress notifications

### CODE REVIEW

File: types/jquery/jquery-tests.ts [hunk 1]
```diff
@@ -3531,6 +3534,58 @@ function JQueryStatic() {
         $();
     }
 
+    function Event() {
+        // #ExpectType Event<HTMLElement>
```
Comment (Kovensky) [inline types/jquery/jquery-tests.ts]: What's the difference between` $ExpectType` and `#ExpectType`

File: types/jquery/jquery-tests.ts [hunk 2]
```diff
@@ -4342,24 +4446,366 @@ function JQueryStatic() {
...
+        function Promise3() {
+            const t = $.ajax() as JQuery.jqXHR<string>;
+            const u = $.ajax() as JQuery.jqXHR<number>;
+            const v = $.ajax() as JQuery.jqXHR<boolean>;
+
+            // 3 parameters
+            {
+                const w = $.when(t, u, v);
+
+                w.then((a, b, c) => {
```
Comment (Kovensky) [inline types/jquery/jquery-tests.ts]: That's not valid in JQuery 3.

JQuery 3 `.then` and `.catch` always only pass exactly one argument to their callback, like the Promises/A+ spec. `.done` and `.fail` should be used instead if multiple arguments are required.

I shipped this bug into production earlier this year :cry:

File: types/jquery/index.d.ts [hunk 3]
```diff
@@ -2389,12 +2390,24 @@ interface JQuery<TElement extends Node = HTMLElement> {
      * @since 1.4
      */
     wrapInner(wrappingElement: JQuery.Selector | JQuery.htmlString | Element | JQuery | ((this: TElement, index: number) => string | JQuery | Element)): this;
+
+    [n: number]: TElement;
 }
 
-interface JQuery<TElement extends Node = HTMLElement> extends ArrayLike<TElement>, Iterable<TElement> { }
+interface JQuery<TElement extends Node = HTMLElement> extends Iterable<TElement> { }
```
Comment (Kovensky) [inline types/jquery/index.d.ts]: Interesting, it really does have a proper `Symbol.iterator`.

It is also `ArrayLike`, right?

File: types/jquery/index.d.ts [hunk 4]
```diff
@@ -3265,9 +3267,103 @@ interface JQueryStatic<TElement extends Node = HTMLElement> {
...
+        UR3 = never, UJ3 = never,
+        VR3 = never, VJ3 = never>
+        (deferredT: JQuery.Promise3<TR1, TJ1, any, TR2, TJ2, any, TR3, TJ3, any> |
+             JQuery.Promise2<TR1, TJ1, any, TR2, TJ2, any>,
+         deferredU: JQuery.Promise3<UR1, UJ1, any, UR2, UJ2, any, UR3, UJ3, any> |
+             JQuery.Promise2<UR1, UJ1, any, UR2, UJ2, any>,
+         deferredV: JQuery.Promise3<VR1, VJ1, any, VR2, VJ2, any, VR3, VJ3, any> |
+             JQuery.Promise2<VR1, VJ1, any, VR2, VJ2, any>): JQuery.Promise3<[TR1, TR2, TR3], [TJ1, TJ2, TJ3], never,
+        [UR1, UR2, UR3], [UJ1, UJ2, UJ3], never,
+        [VR1, VR2, VR3], [VJ1, VJ2, VJ3], never>;
```
Comment (Kovensky) [inline types/jquery/index.d.ts]: 😨

**Review by Jessidhia — COMMENTED** [review]
I completely glossed over the Deferred/Promise section in the definitions, though... dealing with all those type arguments is hard D:

File: types/jquery/jquery-tests.ts (hunk 1 above)
Comment (leonard-thieu) [inline types/jquery/jquery-tests.ts]: oops

I typo'd and apparently copy/pasted it a few times.

File: types/jquery/jquery-tests.ts (hunk 2 above)
Comment (leonard-thieu) [inline types/jquery/jquery-tests.ts]: Seems to provide multiple (https://fiddle.jshell.net/leonard_thieu/36unLssw/). Does that violate the spec?

File: types/jquery/index.d.ts (hunk 3 above)
Comment (leonard-thieu) [inline types/jquery/index.d.ts]: Yes, but `ArrayLike`'s definition has the index signature marked as `readonly`.
```typescript
interface ArrayLike<T> {
    readonly length: number;
    readonly [n: number]: T;
}
```
I added an index signature to `JQuery` to compensate (it already has a length property).

It looks like writing through the index signature is a [real scenario](https://github.com/jquery/sizzle/wiki#sizzleselectorspseudosname--function-elem--).
```javascript
var $test = jQuery( document );
Sizzle.selectors.pseudos.fixed = function( elem ) {
    $test[ 0 ] = elem;
    return $test.css( "position" ) === "fixed";
};
```

File: types/jquery/index.d.ts (hunk 4 above)
Comment (leonard-thieu) [inline types/jquery/index.d.ts]: Same here.

**Comment by leonard-thieu** [comment]
27 type parameters is brutal. I'm hoping variadic types fixes this (and that we get it soon).

**Comment by leonard-thieu** [comment]
I really need to figure out why it stops building on this repo sometimes. See build results at https://github.com/leonard-thieu/DefinitelyTyped/pull/7.

File: types/jquery/jquery-tests.ts (hunk 2 above)
Comment (leonard-thieu) [inline types/jquery/jquery-tests.ts]: Actually, it looks like the spec doesn't preclude more than 1 argument (https://promisesaplus.com/#point-26).

> If onFulfilled is a function:
> - it must be called after promise is fulfilled, with promise’s value as its first argument.

> If onRejected is a function,
> - it must be called after promise is rejected, with promise’s reason as its first argument.

**Comment by leonard-thieu** [comment]
I've been working on the assumption that `pipe()` is an alias to `then()` but that turns out to not be true. `pipe()` will pass the return value of `failFilter` to the next `failFilter`. `then()` will pass the return value of `failFilter` to the next `doneFilter`.

I made this well-drawn illustration as a visual aid.

jquery-promises-1

Fiddle: https://fiddle.jshell.net/leonard_thieu/zmwo0bqp/

From looking at jQuery's tests, there might be more differences that I have to investigate.

**Comment by leonard-thieu** [comment]
I don't understand how this ever worked in the first place.

**Comment by leonard-thieu** [comment]
I have the following two test cases:
```typescript
    function Promise() {
        const p1: JQuery.Promise<I1, I2, I3> = {} as any;
        const p2: JQuery.Promise<I2, I3, I4> = {} as any;
        const p3: JQuery.Promise<I3, I4, I5> = {} as any;

        // 3 parameters
        {
            const w = $.when(p1, p2, p3);

            w.then((a, b, c) => {
                a; // $ExpectType I1
                b; // $ExpectType I2
                c; // $ExpectType I3
            } /* ... */ );
        }
    }
```
```typescript
    // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/10159
    function issue_10159() {
        interface One { result: number; }
        interface Two { Result: number; }
        interface Three { TheResult: number; }

        class AsyncRunner {
            Run(): void {
                const task1: JQuery.Promise<One, any, any> = {} as any;
                const task2: JQuery.Promise<Two, any, any> = this.runTask2();
                const task3: JQuery.Promise<Three, any, any> = this.runTask3();

                $.when(task1, task2, task3)
                    .then((r1, r2, r3) => {
                        r1; // $ExpectType One
                        r2; // $ExpectType Two
                        r3; // $ExpectType Three
                    });
            }

            runTask1() {
                let dfd = $.Deferred<One>();
                console.log('Task 1');
                setTimeout(() => { dfd.resolve({ result: 1 }); }, Math.floor(400 + Math.random() * 2000));
                return dfd.promise();
            }

            runTask2() {
                let dfd = $.Deferred<Two>();
                console.log('Task 2');
                setTimeout(() => { dfd.resolve({ Result: 2 }); }, Math.floor(400 + Math.random() * 2000));
                return dfd.promise();
            }

            runTask3() {
                let dfd = $.Deferred<Three>();
                console.log('Task 3');
                setTimeout(() => { dfd.resolve({ TheResult: 3 }); }, Math.floor(400 + Math.random() * 2000));
                return dfd.promise();
            }
        }
    }
```
Maybe I've been looking at this for too long but shouldn't
```typescript
const w = $.when(p1, p2, p3);

w.then((a, b, c) => {
    a; // $ExpectType I1
    b; // $ExpectType I2
    c; // $ExpectType I3
} /* ... */ );
```
and
```typescript
$.when(task1, task2, task3)
    .then((r1, r2, r3) => {
        r1; // $ExpectType One
        r2; // $ExpectType Two
        r3; // $ExpectType Three
    });
```
behave the same way?

Right now, the 2nd one is failing with
```
ERROR: 4765:25  expect  Expected type to be:
  One
got:
  [One, One, One]
ERROR: 4766:25  expect  Expected type to be:
  Two
got:
  [Two, Two, Two]
ERROR: 4767:25  expect  Expected type to be:
  Three
got:
  [Three, Three, Three]
```

It looks like it's matching with the `Promise3` overload of `when()` instead of the `Promise` overload.

**Comment by leonard-thieu** [comment]
Changing the type parameters for the arguments seems to impact which overload they hit.

```typescript
const task1: JQuery.Promise<One, any, any> = this.runTask1();       // Fail
const task1: JQuery.Promise<One, Two, any> = this.runTask1();       // Pass
const task1: JQuery.Promise<One, any, Two> = this.runTask1();       // Fail
const task1: JQuery.Promise<One, One, any> = this.runTask1();       // Fail
const task1: JQuery.Promise<One, never, any> = this.runTask1();     // Fail
const task1: JQuery.Promise<One, string, any> = this.runTask1();    // Pass
const task1: JQuery.Promise<One, undefined, any> = this.runTask1(); // Fail
```
I don't understand why 😭

**Comment by leonard-thieu** [comment]
Looks like I've managed to break CI on both repos.

I've backed off on the multiple multi-argument Promise overloads. This should be good to go @johnnyreilly

**Comment by johnnyreilly** [comment]
Thanks!

**Pull request merged**